joblib/parallel.py

```diff
@@ -774,65 +793,153 @@ def print_progress(self):
+            # to leave it in a working state to allow for future jobs
+            # scheduling.
+            ensure_ready = self._managed_backend
+            backend.abort_everything(ensure_ready=ensure_ready)
+
+        if not isinstance(exception, TransportableException):
+            raise
+        else:
+            # Capture exception to add information on the local
+            # stack in addition to the distant stack
+            this_report = format_outer_frames(context=10,
+                                              stack_start=1)
+            report = """Multiprocessing exception:
 %s
 ---------------------------------------------------------------------------
 Sub-process traceback:
 ---------------------------------------------------------------------------
 %s""" % (this_report, exception.message)
-                    # Convert this to a JoblibException
-                    exception_type = _mk_exception(exception.etype)[0]
-                    exception = exception_type(report)
+            # Convert this to a JoblibException
+            exception_type = _mk_exception(exception.etype)[0]
+            exception = exception_type(report)
+
+            raise exception
+
+    def _terminate_and_reset(self):
+        if not self._managed_backend:
+            self._terminate_backend()
+        self._jobs = list()
+        self._pending_outputs = list()
+
+    def _start(self, iterator, pre_dispatch, n_jobs):
+        try:
+            # Only set self._iterating to True if at least a batch
+            # was dispatched. In particular this covers the edge
+            # case of Parallel used with an exhausted iterator. If
+            # self._original_iterator is None, then this means either
+            # that pre_dispatch == "all", n_jobs == 1 or that the first batch
+            # was very quick and its callback already dispatched all the
+            # remaining jobs.
+            self._iterating = False
+            if self.dispatch_one_batch(iterator):
+                self._iterating = self._original_iterator is not None
+
+            while self.dispatch_one_batch(iterator):
+                pass
+
+            if pre_dispatch == "all" or n_jobs == 1:
+                # The iterable was consumed all at once by the above for loop.
+                # No need to wait for async callbacks to trigger to
+                # consumption.
+                self._iterating = False
 
-                    raise exception
+        except BaseException as exception:
+            try:
+                self._handle_exception(exception)
+            except BaseException:
+                self._terminate_and_reset()
+                raise
+
+    def _next_output(self):
+        # We need to be careful: the buffer list can be filling
+        # up as we empty it and Python list are not thread-safe
+        # up default hence the use of the lock
+        with self._lock:
+            job = self._jobs.pop(0)
```

Comment:
The `list.pop` operation is thread-safe in python. The list access is protected by the GIL and the `pop` operation is atomic.
The protection with the lock could be removed (unless I am missing something here..)

Reply:
This was already there, I didn't try to remove it